### ISSUE TITLE
XMLHttpRequestUpload events not firing in Expo Go 49+

### ISSUE THREAD

**Issue by cshfang** [expo/expo#28269]
### Minimal reproducible example

https://github.com/cshfang/xhr-issue

### What platform(s) does this occur on?

iOS

### Did you reproduce this issue in a development build?

No (tested in Expo Go)

### Summary

XMLHttpRequestUpload events appear to be not firing in Expo Go 49+

Related discussion found here: https://github.com/expo/expo/discussions/27448

Example code of xhr events not firing:
```js
xhr.upload.addEventListener('progress', event => {
  console.log('Event: Upload progress', event); // does not fire during upload
});
xhr.upload.onprogress = event => {
  console.log('Upload progress', event); // does not fire during upload
};
```

| Platform | XMLHttpRequestUpload events firing |
| :-- | :--: |
| React Native 0.71 | ✅ |
| React Native 0.73 | ✅ |
| Expo Go SDK 48 | ✅ |
| Expo Go SDK 50 | ❌ |
| Expo Dev Build SDK 50 | ✅ |

### Environment

Run within `xhr-latest` (app presenting with issue) directory
```
  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.4
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.18.0 - ~/.nvm/versions/node/v18.18.0/bin/node
      Yarn: 1.22.19 - ~/.nvm/versions/node/v18.18.0/bin/yarn
      npm: 9.8.1 - ~/.nvm/versions/node/v18.18.0/bin/npm
    Managers:
      CocoaPods: 1.13.0 - /Users/chrfang/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.0, iOS 17.0, macOS 14.0, tvOS 17.0, watchOS 10.0
      Android SDK:
        API Levels: 26, 28, 30, 31, 32, 33, 34
        Build Tools: 28.0.3, 30.0.2, 30.0.3, 31.0.0, 33.0.0, 33.0.1, 34.0.0, 34.0.0
        System Images: android-24 | Google APIs ARM 64 v8a, android-26 | Google APIs ARM 64 v8a, android-32 | Google APIs ARM 64 v8a, android-33 | Google APIs ARM 64 v8a, android-33 | Google Play ARM 64 v8a, android-34 | Google Play ARM 64 v8a
    IDEs:
      Android Studio: 2022.3 AI-223.8836.35.2231.10671973
      Xcode: 15.0/15A240d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~50.0.14 => 50.0.17 
      react: 18.2.0 => 18.2.0 
      react-native: 0.73.6 => 0.73.6 
    npmGlobalPackages:
      eas-cli: 5.9.1
    Expo Workflow: managed
```

### Expo Doctor Diagnostics

Run within `xhr-latest` (app presenting with issue) directory
```
✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check npm/ yarn versions
✔ Check for issues with metro config
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that native modules do not use incompatible support packages
✔ Check for legacy global CLI installed locally
✔ Check that native modules use compatible support package versions for installed Expo SDK
✔ Check that packages match versions required by installed Expo SDK

Didn't find any issues with the project!
```

**Comment by ChristopherGabba** [comment]
Referencing other issues:
https://github.com/numandev1/react-native-compressor/issues/276
https://github.com/expo/expo/discussions/27448 

I'm seeing this as well.

**Comment by faarda** [comment]
Did you find a work around for this @cshfang ?

**Comment by brentvatne** [comment]
@ChristopherGabba - as far as i know this only impacts expo go, as reported here, so it should not have any impact on your users. are you experiencing something different?

**Comment by brentvatne** [comment]
this seems to coincide with the release of network debugging: https://blog.expo.dev/expo-sdk-49-c6d398cdf740

i was able to reproduce it in expo go in a new app by copying @cshfang's app code over to a new project. outside of expo go, it works as expected (eg: via `npx expo run:ios`)

```
import React from 'react';
import {Button, StyleSheet, View} from 'react-native';

const url = 'http://127.0.0.1:3000';
export default function App() {
  const upload = () => {
    const xhr = new XMLHttpRequest();
    xhr.open('POST', url);
    xhr.upload.addEventListener('progress', event => {
      console.log('Event: Upload progress', event);
    });
    xhr.upload.onprogress = event => {
      console.log('Upload progress', event);
    };
    xhr.onprogress = event => {
      console.log('Progress', event);
    };
    xhr.onload = event => {
      console.log('Completed', event);
    };
    xhr.send('just a string');
  };
  return (
    <View style={styles.container}>
      <Button title="Upload" onPress={upload} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

//  LOG  Event: Upload progress {"isTrusted": false, "lengthComputable": true, "loaded": 13, "total": 13}
//  LOG  Upload progress {"isTrusted": false, "lengthComputable": true, "loaded": 13, "total": 13}
//  LOG  Progress {"isTrusted": false, "lengthComputable": false, "loaded": 13, "total": -1}
//  LOG  Completed {"isTrusted": false}
```

and then running his provided server code:

```
const http = require('http')

const server = http.createServer(function(request, response) {
  if (request.method == 'POST') {
    var body = ''
    request.on('data', function(data) {
      body += data
    })
    request.on('end', function() {
      console.log('Body: ' + body)
      response.writeHead(200, {'Content-Type': 'text/html'})
      response.end('post received')
    })
  }
})

const port = 3000
const host = '127.0.0.1'
server.listen(port, host)
console.log(`Listening at http://${host}:${port}`)
```

**Comment by ChristopherGabba** [comment]
@brentvatne @Kudo I am experiencing this issue in my production app on Expo 51 as well (on iOS)so it's not just Expo Go. I'm unable to get progress feedback on uploads.

**Comment by brentvatne** [comment]
@ChristopherGabba - we can only help on reports where there is a minimal reproducible example. we are unable to reproduce this outside of expo go. please share a minimal reproducible example in a new issue if you believe that you have found a bug - this issue is about expo go

**Comment by ChristopherGabba** [comment]
@brentvatne I apologize, I'll work on it. @cshfang posted this issue on behalf of the AWS team who were troubleshooting my issue on AWS: https://github.com/aws-amplify/amplify-js/issues/13221

**Comment by brentvatne** [comment]
@ChristopherGabba - if you can reproduce it without using amplify that would be great! the code provided by the OP in this thread works as expected in a release build

**Comment by ChristopherGabba** [comment]
@brentvatne Super strange. The upload progress is actually working on a production TestFlight build, but not on the Expo Dev Build -- just using the basic amplify `uploadData` (https://docs.amplify.aws/react-native/build-a-backend/storage/upload-files/)  function. I coded it into the app, and it wasn't working on Dev and as soon as I downloaded the Testflight build it works awesome -- still can't get any in the development build though.

Regardless, I have not been able to reproduce it without the basic Amplify library, but just providing an update.

**Comment by EtherealData** [comment]
bump? app feels broken with uploads, this has been broken all year.

**Comment by phlcastro** [comment]
@Kudo any updates on this one? I'm curious as very few people complained about it. do people use a different solution to upload files and display the progress? 🤔

**Comment by skjortan23** [comment]
I have the same issue. `onProgress` just will not report progress.
I will test a production build and see if that works. if its only for the dev client then its manageable. But if broken in the production build then its a show stopper.

**Comment by markrickert** [comment]
Hey @brentvatne I just ran into this issue while converting a client's react-native `0.76.6` app to expo sdk. 

Couldn't figure out why the progress bars in my app weren't showing when uploading files using `XMLHttpRequest`. The bare react native app on `0.76.6` properly shows the progress bars so i think you might be right that something in expo sdk 49 network debugging broke this.

I believe that the reproduction examples haven't been getting the criteria quite right... they're just sending [small strings](https://github.com/cshfang/xhr-issue/blob/main/rnxhr71/App.tsx#L21). However I'm using `FormData()` to send very large files to a server. One would expect the `xhr.upload.onprogress` event (or the listener) to fire multiple times during a multi-second upload, however it fires only once at the very end of the network request, which is why i think that the repros might have been giving false positives about this issue.

I've created a minimal reproduction app that has a small express http server that can accept `multipart/form` content types and has it's bandwidth throttled so that the 4mb file we're uploading from the app to `localhost` doesn't just transfer instantly... takes some time so multiple callbacks should happen in the app for progress handling.

Please let me know if i can help in any way to get this resolved!

https://github.com/markrickert/expo-bug-XMLHttpRequest-onprogress

Here's the difference in what i'm seeing in functionality

Expo SDK 48: progress handlers fire correctly (even though it gets the percentage wrong)

![Image](https://github.com/user-attachments/assets/2428cda4-fd8b-42c9-b05e-770570feacaf)

Expo SDK 49+ expo go and dev client all do not fire progress handlers correctly

![Image](https://github.com/user-attachments/assets/d8083ce0-4f01-457b-bd18-15443b5919f3)

**Comment by ChristopherGabba** [comment]
Thanks @Kudo !!

What expo version will this be fixed in? Any specific release?

**Comment by ChristopherGabba** [comment]
@markrickert @Kudo Are you seeing this as fixed on iOS with your repro? I just tried on development build and it just seems like I'm getting no logs on uploadProgress for iOS on development builds still. Expo 53.0.10, react-native 0.79.3

**Comment by JamieGoodson** [comment]
Expo 53 here, iOS development build - not seeing `xhr.upload.onprogress` events fire either.

Weirdly, on Android it works just fine.

**Comment by markrickert** [comment]
I just updated my demo app to sdk 54 and the `xhr.upload.onprogress` events are *still* not firing. 

However, they work fine if i turn `EX_DEV_CLIENT_NETWORK_INSPECTOR` to "false" in the `ios/Podfile.properties.json`.

**Comment by gabimoncha** [comment]
> I just updated my demo app to sdk 54 and the `xhr.upload.onprogress` events are _still_ not firing.
> 
> However, they work fine if i turn `EX_DEV_CLIENT_NETWORK_INSPECTOR` to "false" in the `ios/Podfile.properties.json`.

thanks @markrickert. that did the trick for me too.

another why to set `EX_DEV_CLIENT_NETWORK_INSPECTOR: false` just once, is to set `ios.networkInspector: false` through `expo-build-properties`

cc @Kudo not sure if this should be closed

**Comment by amritbanerjee** [comment]
This makes ios expo go usage unusable still. Is there any updates on this or any workaround for this. Current solution needs dev build.
